Fix sub-circuit padding & Pk/Vk non-determinisim to prevent variadic circuit sizes

CPerezz opened the pull request:
- [x] CopyCircuit: Included `Padding` tag & `max_copy_rows` to introduce correct padding. 
- [x] TxCircuit:  No changes were required. 
- [x] EvmCircuit: No changes required.
- [x] StateCircuit: No changes were required.
- [x] PiCircuit: No changes were required.
- [x] KeccakCircuit: No changes required.


Resolves: #986

CPerezz commented:
`StateCircuit` was also checked and with different witness amounts lead to same circuits.

See the testing code used (requires latest halo2 tag so will not stay within the PR):
<details>

```rust
#[test]
fn variadic_size_check() {
    let mut rows = vec![
        Rw::Stack {
            rw_counter: 24,
            is_write: true,
            call_id: 1,
            stack_pointer: 1022,
            value: U256::from(394500u64),
        },
        Rw::Stack {
            rw_counter: 25,
            is_write: false,
            call_id: 1,
            stack_pointer: 1022,
            value: U256::from(394500u64),
        },
    ];

    let updates = MptUpdates::mock_from(&rows);
    let circuit = StateCircuit::<Fr> {
        rows: rows.clone(),
        updates,
        overrides: HashMap::default(),
        n_rows: N_ROWS,
        _marker: std::marker::PhantomData::default(),
    };
    let power_of_randomness = circuit.instance();
    let prover1 = MockProver::<Fr>::run(17, &circuit, power_of_randomness).unwrap();

    rows.extend_from_slice(&[
        Rw::Stack {
            rw_counter: 26,
            is_write: true,
            call_id: 1,
            stack_pointer: 1021,
            value: U256::from(394511u64),
        },
        Rw::Stack {
            rw_counter: 27,
            is_write: false,
            call_id: 1,
            stack_pointer: 1021,
            value: U256::from(394511u64),
        },
    ]);

    let updates = MptUpdates::mock_from(&rows);
    let circuit = StateCircuit::<Fr> {
        rows,
        updates,
        overrides: HashMap::default(),
        n_rows: N_ROWS,
        _marker: std::marker::PhantomData::default(),
    };
    let power_of_randomness = circuit.instance();
    let prover2 = MockProver::<Fr>::run(17, &circuit, power_of_randomness).unwrap();

    assert_eq!(prover1.fixed(), prover2.fixed());
    assert_eq!(prover1.permutation(), prover2.permutation());
}
```

</details>

CPerezz commented:
For `PiCircuit`, we've increased the amount of Txs witnessed from 4 to 8. No changes have happened to the circuit.
The test applied was:
<details>

```rust
#[test]
    fn variadic_size_check() {
        const MAX_TXS: usize = 8;
        const MAX_CALLDATA: usize = 200;

        let mut rng = ChaCha20Rng::seed_from_u64(2);
        let randomness = Fr::random(&mut rng);
        let rand_rpi = Fr::random(&mut rng);

        let mut public_data = PublicData::default();
        let chain_id = 1337u64;
        public_data.chain_id = Word::from(chain_id);

        let n_tx = 4;
        for i in 0..n_tx {
            let eth_tx = eth_types::Transaction::from(&rand_tx(&mut rng, chain_id, i & 2 == 0));
            public_data.transactions.push(eth_tx);
        }
        let circuit = PiTestCircuit::<Fr, MAX_TXS, MAX_CALLDATA>(PiCircuit::new(
            MAX_TXS,
            MAX_CALLDATA,
            randomness,
            rand_rpi,
            public_data.clone(),
        ));
        let public_inputs = circuit.0.instance();
        let prover1 = MockProver::run(20, &circuit, public_inputs).unwrap();

        // Push 4 more txs
        for i in 0..n_tx {
            let eth_tx = eth_types::Transaction::from(&rand_tx(&mut rng, chain_id, i & 2 == 0));
            public_data.transactions.push(eth_tx);
        }
        let circuit2 = PiTestCircuit::<Fr, MAX_TXS, MAX_CALLDATA>(PiCircuit::new(
            MAX_TXS,
            MAX_CALLDATA,
            randomness,
            rand_rpi,
            public_data,
        ));
        let public_inputs = circuit2.0.instance();
        let prover2 = MockProver::run(20, &circuit, public_inputs).unwrap();

        assert_eq!(prover1.fixed(), prover2.fixed());
        assert_eq!(prover1.permutation(), prover2.permutation());
    }
```

</details?

CPerezz commented:
As with `PiCircuit`, in `TxCircuit` we increase the number of witnessed txs so that we have more signatures to verify.
No issues arised.

See the testing done here:
<details>

```rust
#[test]
    fn variadic_size_check() {
        const MAX_TXS: usize = 2;
        const MAX_CALLDATA: usize = 32;

        let chain_id: u64 = mock::MOCK_CHAIN_ID.as_u64();
        let tx1: Transaction = mock::CORRECT_MOCK_TXS[0].clone().into();
        let tx2: Transaction = mock::CORRECT_MOCK_TXS[1].clone().into();
        let circuit = TxCircuit::<Fr>::new(MAX_TXS, MAX_CALLDATA, chain_id, vec![tx1.clone()]);
        let prover1 = MockProver::<Fr>::run(20, &circuit, vec![vec![]]).unwrap();

        let circuit = TxCircuit::<Fr>::new(MAX_TXS, MAX_CALLDATA, chain_id, vec![tx1, tx2]);
        let prover2 = MockProver::<Fr>::run(20, &circuit, vec![vec![]]).unwrap();

        assert_eq!(prover1.fixed(), prover2.fixed());
        assert_eq!(prover1.permutation(), prover2.permutation());
    }
```

</details?

ed255 commented:
For `KeccakCircuit` I've written this tests, which runs a circuit with empty input, and another one with some hash inputs of different lengths.  The test passes

The test is this one:
<details>

```rust
    #[test]
    fn variadic_size_check() {
        let k = 11;
        let num_rows = Some(2usize.pow(k));
        // Empty
        let inputs = vec![];
        let circuit = KeccakCircuit::new(num_rows, inputs);
        let prover1 = MockProver::<Fr>::run(k, &circuit, vec![]).unwrap();

        // Non-empty
        let inputs = vec![
            vec![],
            (0u8..1).collect::<Vec<_>>(),
            (0u8..135).collect::<Vec<_>>(),
            (0u8..136).collect::<Vec<_>>(),
            (0u8..200).collect::<Vec<_>>(),
        ];
        let circuit = KeccakCircuit::new(num_rows, inputs);
        let prover2 = MockProver::<Fr>::run(k, &circuit, vec![]).unwrap();

        assert_eq!(prover1.fixed(), prover2.fixed());
        assert_eq!(prover1.permutation(), prover2.permutation());
    }
```

CPerezz commented:
`EvmCircuit` works correctly too independently of the ammount of opcodes we pass to it.
See this testing code done for it:

<details>

```rust
fn test_ok(opcode: OpcodeId, bytes: &[u8]) -> MockProver<Fr> {
        assert!(bytes.len() == opcode.data_len());

        let mut bytecode = bytecode! {
            .write_op(opcode)
        };
        for b in bytes {
            bytecode.write(*b, false);
        }
        bytecode.write_op(OpcodeId::STOP);

        let test_ctx = TestContext::<2, 1>::simple_ctx_with_bytecode(bytecode).unwrap();
        let block: GethData = test_ctx.into();
        let mut builder =
            BlockData::new_from_geth_data_with_params(block.clone(), CircuitsParams::default())
                .new_circuit_input_builder();
        builder
            .handle_block(&block.eth_block, &block.geth_traces)
            .unwrap();

        // build a witness block from trace result
        let block = crate::witness::block_convert::<Fr>(&builder.block, &builder.code_db).unwrap();
        let k = get_test_degree(&block);

        let circuit = get_test_cicuit_from_block(block);
        MockProver::<Fr>::run(k, &circuit, vec![]).unwrap()
    }

#[test]
    fn push_gadget_simple() {
        let prover = test_ok(OpcodeId::PUSH1, &[1]);
        let prover2 = test_ok(OpcodeId::PUSH2, &[1, 2]);
        let prover3 = test_ok(
            OpcodeId::PUSH31,
            &[
                1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
                24, 25, 26, 27, 28, 29, 30, 31,
            ],
        );
        let prover4 = test_ok(
            OpcodeId::PUSH32,
            &[
                1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
                24, 25, 26, 27, 28, 29, 30, 31, 32,
            ],
        );

        assert!(
            prover.fixed() == prover2.fixed()
                && prover2.fixed() == prover3.fixed()
                && prover3.fixed() == prover4.fixed()
        );

        assert!(
            prover.permutation() == prover2.permutation()
                && prover2.permutation() == prover3.permutation()
                && prover3.permutation() == prover4.permutation()
        );
    }
```

</details>

han0110 commented on zkevm-circuits/src/copy_circuit.rs:
```diff
@@ -553,7 +588,7 @@ impl<F: Field> CopyCircuitConfig<F> {
             self.copy_table.src_addr_end,
             offset,
-            || Value::known(F::zero()),
+            || Value::known(F::one()),
```
I was confusing in the beginning to see this updated to `1`, then I realized it's required for passing `is_pad` constraint `is_pad == 1 - (src_addr < src_addr_end) for read row`.

I think setting `is_pad` to be `(offset % 2 == 0)` directly makes it easier to follow for me, but it's fine to keep it unchanged (since it doesn't affect performance).

han0110 reviewed (commented):
LGTM! Nice work! Only left some nitpicks but not necessary to address.

For those failing tests, I just tried to set the correct param and they are also working:

- `copy_circuit::tests::copy_circuit_valid_calldatacopy`
    https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b433af45c9dabfc1c938bb0fd2cf671e3e3c8c2b/zkevm-circuits/src/copy_circuit.rs#L838-L845

    Need a diff like:

    ```diff
             let mut builder = BlockData::new_from_geth_data_with_params(
                 block.clone(),
                 CircuitsParams {
                     max_rws: 8192,
    +                max_copy_rows: 8192 + 2,
                     ..Default::default()
                 },
             )
             .new_circuit_input_builder();
    ```
    
- `copy_circuit::tests::copy_circuit_valid_sha3`
    https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b433af45c9dabfc1c938bb0fd2cf671e3e3c8c2b/zkevm-circuits/src/copy_circuit.rs#L909-L916

    Need a diff like:

    ```diff
             let mut builder = BlockData::new_from_geth_data_with_params(
                 block.clone(),
                 CircuitsParams {
                     max_rws: 2000,
    +                max_copy_rows: 0x200 * 2 + 2,
                     ..Default::default()
                 },
             )
             .new_circuit_input_builder();
    ```

CPerezz commented on bus-mapping/src/circuit_input_builder/execution.rs:
```diff
@@ -155,8 +155,11 @@ impl ExecState {
 pub enum CopyDataType {
+    /// When we need to pad the Copy rows of the circuit up to a certain maximum
+    /// with rows that are not "useful".
+    Padding = 0,
```
It made sense to @ed255 and me as the default behaviour is to copy from Memory. But we can revert it. Whatever you prefer.

ed255 reviewed (approved):
LGTM!  Just one note, after this PR is merged we will still be missing the ExpCircuit (which should be addressed via https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1062 ) so let's keep the related issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/986 open after we merge this :)
